contracts/RemovalQueue.sol

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.13;
+import {RemovalUtils} from "./RemovalUtils.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlEnumerableUpgradeable.sol";
+import "./Removal.sol";
+
+struct RemovalQueueByVintage {
+  mapping(uint16 => EnumerableSetUpgradeable.UintSet) queueByVintage;
+  uint16 earliestYear;
+  uint16 latestYear;
+}
+
+library RemovalQueue {
+  using RemovalUtils for uint256;
+  using EnumerableSetUpgradeable for EnumerableSetUpgradeable.UintSet;
+
+  uint16 private constant _DEFAULT_EARLIEST_YEAR = 2**16 - 1;
+  uint16 private constant _DEFAULT_LATEST_YEAR = 0;
+
+  /**
+   * @notice Inserts a new removal to the queue.
+   * @dev The removal is added to the Enumberable Set that maps to the year of its vintage.
+   * @param removalToInsert a new removal to insert.
+   * @return bool true if success, false otherwise.
+   */
+  function insertRemovalByVintage(
+    RemovalQueueByVintage storage removalQueue,
+    uint256 removalToInsert
+  ) internal returns (bool) {
+    uint16 vintageOfRemoval = removalToInsert.vintage();
+    if (isRemovalQueueEmpty(removalQueue)) {
+      removalQueue.earliestYear = vintageOfRemoval;
+      removalQueue.latestYear = vintageOfRemoval;
+    } else if (vintageOfRemoval < removalQueue.earliestYear) {
+      removalQueue.earliestYear = vintageOfRemoval;
+    } else if (vintageOfRemoval > removalQueue.latestYear) {
+      removalQueue.latestYear = vintageOfRemoval;
+    }
+    return removalQueue.queueByVintage[vintageOfRemoval].add(removalToInsert);
+  }
+
+  /**
+   * @notice Removes a removal from the queue.
+   * @dev Removes the removal from the Enumerable Set that corresponds to its vintage.
+   * @param removalQueue the queue to search through.
+   * @param removalToRemove the removal to remove.
+   * @return bool true if success, false otherwise.
+   */
+  function removeRemoval(
+    RemovalQueueByVintage storage removalQueue,
+    uint256 removalToRemove
+  ) internal returns (bool) {
+    uint16 vintageOfRemoval = removalToRemove.vintage();
+    require(
+      removalQueue.queueByVintage[vintageOfRemoval].remove(removalToRemove),
+      "Market: failed to remove correct removal"
+    );
+    // If all removals were removed, check to see if there are any updates to the struct we need to make.
+    if (isRemovalQueueEmptyForVintage(removalQueue, vintageOfRemoval)) {
+      if (removalQueue.earliestYear == removalQueue.latestYear) {
+        // If there was only one year remaining, clear the values for latest and earliest years.
+        removalQueue.earliestYear = _DEFAULT_EARLIEST_YEAR;
```

Comment:
what are the implications here if the default earliest year is a year in which no future supplier will ever list a removal for again? (e.g., a couple years down the road)

Reply:
The default earliest year is just a way of nulling out this value in such a way where the conditionals will work elsewhere.
Here's the definition:
`uint16 private constant _DEFAULT_EARLIEST_YEAR = 2**16 - 1;`
So I'm not sure I understand your question.

Reply:
if the earliest year is a year far in the past, does thath have any implications on the overhead of iterations?

Reply:
No, since `earliestYear` is always update from the `_DEFAULT_EARLIEST_YEAR` to be whatever was added by a new removal if the default is detected.